### PR TITLE
Disable XSLT on UAP (Ref.Emit not available)

### PR DESCRIPTION
**Pull request by krwq**
Fixes: https://github.com/dotnet/corefx/issues/20739

There are improvements which could happen here:
- Much more code could possibly be removed on UAP: xslt runtime and others - this will require more digging - not sure if it is worth to spend time on it - are we planning to eventually support Ref.Emit on UAP in the future?
- more PNSE could be thrown - user will get PNSE when they try to use it anyway so not sure if this is worth fixing

cc: @danmosemsft

### CODE REVIEW

**Comment by danmosemsft** [comment]
Can you explain how you decided to not fall back to XslTransform on UAP? Is it not sufficiently compatible?

@jkotas

**Comment by krwq** [comment]
@danmosemsft XslTransform has security issues which XslCompiledTransform fixes and also some implementation details might differ - I think it would be better to invest time in making Ref.Emit work and recommend to use old implementation.

https://msdn.microsoft.com/en-us/library/66f54faw.aspx

**Comment by krwq** [comment]
@dotnet-bot test Innerloop netfx Windows_NT Debug x86 Build and Test please

**Comment by jkotas** [comment]
Also, the old XslTransform has known bugs and conformance issues.

**Comment by danmosemsft** [comment]
Of course, if we're driving folks away to XslTransform (because it's the common denominator), we're making the experience worse on all platforms, aren't we?

@jkotas what do you see the longer term path forward -- is Ref.Emit ever likely on UWP, or would we invest in improving XslTransform to the point where we can pass through to it (or conversely add an interpreter for XslCompiledTransform).

**Comment by jkotas** [comment]
> Ref.Emit ever likely on UWP

I do not see anything fundamental that would prevent us from doing it. https://github.com/dotnet/corefx/issues/4491#issuecomment-189756092 would be part of this. 

BTW: Xamarin has added (or is in process of adding) Ref.Emit support for full AOT via IL interpreter. I do not think we would want to be in a situation where Ref.Emit is available everywhere except UWP.

**Comment by jkotas** [comment]
> would we invest in improving XslTransform

I do not think that we would want to invest in improving XslTransform. The reason why XslCompiledTransform was introduced back in the day was that XslTransform was too broken and buggy to be evolved anywhere.

> conversely add an interpreter for XslCompiledTransform

That sounds more reasonable.

**Comment by krwq** [comment]
@jkotas @danmosemsft is this something we want to do as part of this PR? I do not have much context on how much work it is. I think this PR would be an intermediate step either way so not sure if we should extend it much more.

@jkotas how do we currently do aot compilation? do we translate to c++? We could make Ref.Emit work - there are couple of options like KaleidoscopeJIT or symbolic execution. Not sure how much work that would involve - are there any particular reasons we don't want Ref.Emit? (i.e. security)

**Comment by jkotas** [comment]
> is this something we want to do as part of this PR? 

I do not think so.

> how do we currently do aot compilation? do 

We translate IL to native code.

> We could make Ref.Emit work 

Right, there are several options how you can make Ref.Emit work, even on no-JIT AOT platforms.

**Comment by danmosemsft** [comment]
Sounds good. I'm fine with this throwing approach. @jkotas I assume you're fine with non-AOT not throwing? This would be an obvious case where the user could write a bunch of code that works perfectly on F5 but fails when natively compiled.

**Comment by krwq** [comment]
@danmosemsft good catch we might want to fix that just for the sake of confusion 😉

**Comment by danmosemsft** [comment]
@krwq that's one argument, but in other cases we've removed "intentional brokenness" that we had previously added to the F5 code to make it match the AOT behavior. One argument for removing that is we never get the two to match exactly. I'm OK either way, I'm interested in @jkotas 's call.

**Comment by jkotas** [comment]
We have two special builds: one for both uap and one for uapaot.

I think it would be better to have just one special build for uap that also works for uapaot. It make the behavior same between uap and uapaot as side-effect.

**Comment by danmosemsft** [comment]
@krwq I believ eyou just have to change uapaot to uap in your defines and project.

**Comment by danmosemsft** [comment]
you need to fix the tests also :)

**Comment by krwq** [comment]
@danmosemsft we got separate test projects for each feature in xml - I have disabled XslCompiler tests by changing configuration: netstandard => netcoreapp+netfx. I'm running all uap tests now to make sure I disabled them all (there are 3 xslt projects, one was already disabled, i have disabled one and 3rd might need more love too)

**Comment by krwq** [comment]
@danmosemsft System.Xml.Xsl.XslTransformApi.Tests.csproj is hanging when running tests with TargetGroup=uap - before and after this change. I'll leave it since it is irrelevant to my changes (tests for old xslt code)

File: src/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/Configurations.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <BuildConfigurations>
-      netstandard;
+      netfx;
```
Comment (joperezr) [inline src/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/Configurations.props]: Did this not require any changes on the .csproj? I would expect at minimum the Configuration names should have changed. Please address it in your next PR.

File: src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj [hunk 2]
```diff
@@ -4,12 +4,12 @@
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);ReflectionOnly</DefineConstants>
     <ProjectGuid>{4050F1D1-1DD2-4B48-A17B-E3F955518C4B}</ProjectGuid>
-    <DefineConstants Condition="'$(TargetGroup)' == 'uapaot'">$(DefineConstants);uapaot</DefineConstants>
+    <DefineConstants Condition="'$(TargetGroup)' == 'uap'">$(DefineConstants);uap</DefineConstants>
```
Comment (joperezr) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: In here, looks like you are not even using this constant at all so no need to define it, so when you fix the other thing I pointed out, please remove this line.

File: src/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/Configurations.props (hunk 1 above)
Comment (krwq) [inline src/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/Configurations.props]: possibly, I haven't run the script to update

Comment (joperezr) [inline src/System.Private.Xml/tests/Xslt/XslCompiledTransformApi/Configurations.props]: script doesn't work, it's broken :( 
you will have to manually make the change.

**Pull request merged**